### ISSUE TITLE
ZMQ 4.0.6 Breaks Salt

### ISSUE THREAD

**Issue by rallytime** [saltstack/salt#24630]
The newest version of ZMQ that was packaged and released by Arch this morning completely breaks Salt. Stacktraces are below, but the minions aren't able to connect to the master.

**Comment by aboe76** [comment]
I can confirm this on my archlinux box,

**Comment by rallytime** [comment]
I discovered this in the Jenkins Test Suite setup. When the tests start running and salt spins up the master and minion test daemons, the minion daemons cannot authenticate with the master daemon.

Here's the stacktrace on 2015.5:
```
15:18:42  * Setting up Salt daemons to execute tests
15:18:42 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15:25:43 Traceback (most recent call last):
15:25:43   File "/testing/tests/runtests.py", line 476, in <module>
15:25:43     main()
15:25:43   File "/testing/tests/runtests.py", line 461, in main
15:25:43     status = parser.run_integration_tests()
15:25:43   File "/testing/tests/runtests.py", line 379, in run_integration_tests
15:25:43     with TestDaemon(self):
15:25:43   File "/testing/tests/integration/__init__.py", line 184, in __enter__
15:25:43     self.start_zeromq_daemons()
15:25:43   File "/testing/tests/integration/__init__.py", line 240, in start_zeromq_daemons
15:25:43     minion = salt.minion.Minion(self.minion_opts)
15:25:43   File "/testing/salt/minion.py", line 682, in __init__
15:25:43     safe)
15:25:43   File "/testing/salt/minion.py", line 868, in eval_master
15:25:43     if self.authenticate(timeout, safe) == 'full':
15:25:43   File "/testing/salt/minion.py", line 1536, in authenticate
15:25:43     auth = salt.crypt.SAuth(self.opts)
15:25:43   File "/testing/salt/crypt.py", line 269, in __new__
15:25:43     SAuth.instances[key].__singleton_init__(opts)
15:25:43   File "/testing/salt/crypt.py", line 308, in __singleton_init__
15:25:43     self.authenticate()
15:25:43   File "/testing/salt/crypt.py", line 326, in authenticate
15:25:43     creds = self.sign_in(timeout, safe)
15:25:43   File "/testing/salt/crypt.py", line 693, in sign_in
15:25:43     raise SaltClientError('Attempt to authenticate with the salt master failed')
15:25:43 salt.exceptions.SaltClientError: Attempt to authenticate with the salt master failed
15:26:22 Build was aborted
```

And here's the stack track on develop:
```
* [Quit in 0:01:00] Waiting for minion, sub_minionProcess Process-3:
15:18:32 Traceback (most recent call last):
15:18:32   File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
15:18:32     self.run()
15:18:32   File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
15:18:32     self._target(*self._args, **self._kwargs)
15:18:32   File "/testing/tests/integration/__init__.py", line 236, in start
15:18:32     getattr(daemon, start_fun)()
15:18:32   File "/testing/salt/minion.py", line 1594, in tune_in
15:18:32     self.sync_connect_master()
15:18:32   File "/testing/salt/minion.py", line 589, in sync_connect_master
15:18:32     raise six.reraise(*future_exception)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/minion.py", line 596, in connect_master
15:18:32     master, self.pub_channel = yield self.eval_master(self.opts, self.timeout, self.safe)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/minion.py", line 802, in eval_master
15:18:32     yield pub_channel.connect()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/transport/zeromq.py", line 310, in connect
15:18:32     yield self.auth.authenticate()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "<string>", line 3, in raise_exc_info
15:18:32 SaltClientError: Attempt to authenticate with the salt master failed
15:18:32 Process Process-2:
15:18:32 Traceback (most recent call last):
15:18:32   File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
15:18:32     self.run()
15:18:32   File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
15:18:32     self._target(*self._args, **self._kwargs)
15:18:32   File "/testing/tests/integration/__init__.py", line 236, in start
15:18:32     getattr(daemon, start_fun)()
15:18:32   File "/testing/salt/minion.py", line 1594, in tune_in
15:18:32     self.sync_connect_master()
15:18:32   File "/testing/salt/minion.py", line 589, in sync_connect_master
15:18:32     raise six.reraise(*future_exception)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/minion.py", line 596, in connect_master
15:18:32     master, self.pub_channel = yield self.eval_master(self.opts, self.timeout, self.safe)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/minion.py", line 802, in eval_master
15:18:32     yield pub_channel.connect()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 876, in run
15:18:32     yielded = self.gen.throw(*exc_info)
15:18:32   File "/testing/salt/transport/zeromq.py", line 310, in connect
15:18:32     yield self.auth.authenticate()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 870, in run
15:18:32     value = future.result()
15:18:32   File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
15:18:32     raise_exc_info(self._exc_info)
15:18:32   File "<string>", line 3, in raise_exc_info
15:18:32 SaltClientError: Attempt to authenticate with the salt master failed
15:19:32 
                                                                                
 * [Quit in 0:00:57] Waiting for minion, sub_minion

<snipped>

15:19:32  * WARNING: Minions failed to connect back. Tests requiring them WILL fail
15:19:32 ================================================================================
15:19:32 
15:19:32  * ERROR: Minions failed to connect
15:19:32  * Initializing SSH subsystem
15:19:32 ~~~~~~~ Versions Report ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15:19:32 Salt Version:
15:19:32            Salt: 2015.5.0-1511-g9f24ec2
15:19:32  
15:19:32 Dependency Versions:
15:19:32          Jinja2: 2.7.3
15:19:32        M2Crypto: 0.22
15:19:32            Mako: Not Installed
15:19:32          PyYAML: 3.11
15:19:32           PyZMQ: 14.6.0
15:19:32          Python: 2.7.8 (default, Sep 24 2014, 18:26:21)
15:19:32            RAET: Not Installed
15:19:32             ZMQ: 4.0.6
15:19:32           ioflo: Not Installed
15:19:32         libnacl: Not Installed
15:19:32    msgpack-pure: Not Installed
15:19:32  msgpack-python: 0.4.6
15:19:32        pycrypto: 2.6.1
15:19:32  
15:19:32 System Versions:
15:19:32            dist:   
15:19:32         machine: x86_64
15:19:32         release: 4.0.4-x86_64-linode57
15:19:32  
15:19:32 ~~~~~~~ Minion Grains Information~~~~~~~~~~~~~~~~~~~~~~~~~~~
15:19:35 Traceback (most recent call last):
15:19:35   File "/testing/tests/runtests.py", line 476, in <module>
15:19:35     main()
15:19:35   File "/testing/tests/runtests.py", line 461, in main
15:19:35     status = parser.run_integration_tests()
15:19:35   File "/testing/tests/runtests.py", line 379, in run_integration_tests
15:19:35     with TestDaemon(self):
15:19:35   File "/testing/tests/integration/__init__.py", line 214, in __enter__
15:19:35     grains = self.client.cmd('minion', 'grains.items')
15:19:35   File "/testing/salt/client/__init__.py", line 538, in cmd
15:19:35     **kwargs)
15:19:35   File "/testing/salt/client/__init__.py", line 293, in run_job
15:19:35     raise SaltClientError(general_exception)
```

**Comment by cachedout** [comment]
Also confirmed on Arch Linux. Via `tcpdump` we are seeing the minion send a public key over the wire during startup but no reply from the master at all.

**Comment by aboe76** [comment]
on the same system where zeromq fails, raet connects but has some strange issues, with targetting 
I see this in the debug logging, I don't think the 'u' needs to be there..
```
[INFO    ] Fetching file from saltenv 'base', ** done ** u'salt/minion.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** u'salt/map.jinja'
[INFO    ] Fetching file from saltenv 'base', ** done ** u'salt/defaults.yaml'
[INFO    ] Fetching file from saltenv 'base', ** done ** u'ntp/ng/init.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** u'ntp/ng/map.jinja'
```

**Comment by aneeshusa** [comment]
FYI, I've just updated my Arch and Arch ARM systems to zeromq 4.1.1-1 (released on the 13th) and the minions appear to be connecting just fine to the master, I can run commands, etc.

**Comment by thedrow** [comment]
@aboe76 It's nothing serious. u'string' just represents a unicode literal.
Some object is probably not converted to string correctly.

**Comment by cachedout** [comment]
Arch has released 4.1 which works correctly. Discussion is here:

https://bugs.archlinux.org/task/45316?string=zeromq&project=1&search_name=&type[0]=&sev[0]=&pri[0]=&due[0]=&reported[0]=&cat[0]=&status[0]=open&percent[0]=&opened=&dev=&closed=&duedatefrom=&duedateto=&changedfrom=&changedto=&openedfrom=&openedto=&closedfrom=&closedto=
